### PR TITLE
Passthrough EXIF properties while scaling image

### PR DESCRIPTION
**Pull request by hashamyounis9**
/claim #2689

## Product Description

This PR ensures that EXIF metadata (like orientation, timestamp, GPS, etc.) is preserved when an image is scaled down in the CommCare Android app. This improves user experience in workflows that rely on image metadata.

## Technical Summary
[Passthrough EXIF metadata while scaling images](https://github.com/dimagi/commcare-android/issues/2689)

Modified the image scaling logic to preserve EXIF metadata by copying it from the original image to the scaled-down version as can be seen in the following screenshot

Screenshot from 2025-04-06 18-04-38

Values on the left side of `->` are coming from original image and those on the right side are coming from the scaled image.

## Feature Flag
Not behind a feature flag.

### Safety story
- The changes were tested locally with various images containing different EXIF metadata.
- Verified that the metadata is retained correctly after scaling, and that existing image handling workflows remain unaffected.
- The change is safe because it does not modify the core image input/output logic, only appends metadata after scaling.

### Automated test coverage
- No new unit tests added due to the nature of the change (image files), but tested locally with debug logs to verify correct metadata retention.

### QA Plan
- A manual QA test is recommended using images with EXIF metadata (e.g., orientation, GPS location).
- Verify that the metadata is preserved in the scaled version.
- No regression expected in image capture or submission workflows.

## Labels and Review

- [x] Do we need to enhance the manual QA test coverage ? If yes, the "QA Note" label is set correctly
- [ ] Does the PR introduce any major changes worth communicating ? If yes, the "Release Note" label is set and a "Release Note" is specified in PR description.
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by damagatchi** [comment]
Can one of the admins verify this patch?

**Comment by shubham1g5** [comment]
@damagatchi  ok to test

**Review by shubham1g5 — CHANGES_REQUESTED** [review]
Hello, thanks for submitting a PR.

Unfortunately our Jenkins Jobs are restricted, but this is the error we are getting for you reference - 

````
FileUtil.java:10: error: package androidx.exifinterface.media does not exist
````

Also in terms of testing this, can you make sure all these following cases are working correctly - 


1. Basic EXIF Check

Snap a pic with GPS turned on.
Save the form.
Check that all EXIF data is intact using your favorite EXIF viewer.

2. Scaled Image Test

Take a high-res image.
Verify the EXIF data after it gets scaled down.

3. Encrypted Image Test

Capture a GPS-tagged image.
Confirm EXIF data survived the encryption.

4. Edge Cases

Use images with no EXIF data.
Test images from various sources (camera, gallery, etc.).
Try different formats (JPG, PNG).

**Comment by hashamyounis9** [comment]
@shubham1g5 

2c17817c7f0c9a4941fd1646a5bfe7c146515d1d resolves the package not found issue, updates the EXIF_TAGS list as asked, also retains the default functionality of keep going if the exif properties were not read.

Here is what we are capturing now with

GPS on:
gps-on

GPS off:
gps-off

**Comment by hashamyounis9** [comment]
@shubham1g5 

also can you please update me with error in jenkins job?

**Comment by hashamyounis9** [comment]
> thanks @hashamyounis9 , can you confirm the test cases I outlined above as well ?

Sure, I'll do in a bit...

**Comment by hashamyounis9** [comment]
@shubham1g5 

confirming the mentioned cases, it works well in the situations mentioned above, here are some catches 

- PNG format does not support EXIF metadata, so it will return nulls with PNG images, but that isn't a problem because android phones by default use jpg format.

- Also when choosing images from gallery, some of the GPS attributes are not provided as it is most probably due to restriction of android allowing for media and restricts some properties while sharing data (BTW this is out of the scope of this issue/PR as it is only responsible for copying available EXIF properties from original image to scaled one.)

**Comment by hashamyounis9** [comment]
> Got it, that makes sense to me. However we must make sure that the exif data survives the [image encryption](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/views/widgets/MediaWidget.java#L289). Have you been able to check on that ?

Not yet, I'll need to take a loot on that...

**Comment by hashamyounis9** [comment]
@shubham1g5 

can you guide me when does the encryption trigger, here is what I did:

 I started filling the form and captured the image using camera and completed the form and form got sent to the server. During all this encryption didn't occur as I did not the debug logs that I put in the [encryption method](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/views/widgets/MediaWidget.java#L289).

**Comment by hashamyounis9** [comment]
@shubham1g5 

For the sake of testing if the EXIF metadata was preserved or not in the images, I exported the form media from the [CommCare HQ](https://www.commcarehq.org/) for my app sent by mobile worker and viewed the EXIF metadata using some online tool making sure that all the EXIF metadata was present in the images that were sent to the server.

**Comment by hashamyounis9** [comment]
> Ahh Sorry, that might have been incorrect path, Curious what does [this check](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/activities/components/ImageCaptureProcessing.java#L67) returns for you ? And if `true` does the [underlining encryption method](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/activities/components/ImageCaptureProcessing.java#L70) gets called for you ?

So I've checked it and the check returned true for me and encryption method gets called, which essentially means that encryption is getting applied and the EXIF metadata survives the encryption, as the images exported from the [CommCare HQ](https://www.commcarehq.org/) contains the retained EXIF metadata!

**Comment by shubham1g5** [comment]
@hashamyounis9 that's great, Is it possible for you to provide a commcare hq link to one such form with image here. Thanks!

@damagatchi retest this please

**Comment by shubham1g5** [comment]
> you mean qr code for the app?

no I meant the link to the form itself where you uploaded the image.

**Comment by hashamyounis9** [comment]
> > you mean qr code for the app?
> 
> no I meant the link to the form itself where you uploaded the image.

oh, can you guide me how to share this?

**Comment by shubham1g5** [comment]
@damagatchi retest this please

**Comment by shubham1g5** [comment]
@hashamyounis9 Can you provide the image to the final image you are seeing on HQ. You can just take the link from the browser address bar where you see the image and paste it here or upload the image here itself or on Google drive, somewhere I can take the image from and verify the exif attributes on my end.

**Comment by hashamyounis9** [comment]
> @hashamyounis9 Can you provide the link to the final image you are seeing on HQ. You can just take the link from the browser address bar where you see the image and paste it here or upload the image here itself or on Google drive, somewhere I can take the image from and verify the exif attributes on my end.

[Here](https://drive.google.com/file/d/1bFUh8co4dwv9MrGzDXM8YPTL7xDbimfL/view?usp=sharing) is the zip exported from CommCare HQ, with multimedia only

I used [this tool](https://jimpl.com/) to view exif metadata

**Comment by shubham1g5** [comment]
@hashamyounis9 thanks! let me verify it. Also we are facing an unrealted error with our CI job and I am looking into it. 
@damagatchi retest this please

File: app/src/org/commcare/utils/FileUtil.java [hunk 1]
```diff
@@ -609,16 +655,33 @@ public static boolean scaleAndSaveImage(File originalImage, String finalFilePath
...
         if (scaledBitmap != null) {
             // Write this scaled bitmap to the final file location
             try {
-                writeBitmapToDiskAndCleanupHandles(scaledBitmap, type, new File(finalFilePath));
+                writeBitmapToDiskAndCleanupHandles(scaledBitmap, type, scaledFile);
+                ExifInterface newExif = new ExifInterface(scaledFile.getAbsolutePath());
+                copyExifData(originalExif, newExif, scaledBitmap);
+                if (BuildConfig.DEBUG) {
+                    logExifAttributesToCheckRetention(originalExif, newExif);
+                }
```
Comment (shubham1g5) [inline app/src/org/commcare/utils/FileUtil.java:683]: nit: think we are ok to not log this instead.

**Review by shubham1g5 — APPROVED** [review]
one minor omition request, but all things look good otherwise.

**Comment by hashamyounis9** [comment]
@shubham1g5 

b9dc391f10ade89d85d07f82ec6765a3c03be551 removes the logs

File: app/src/org/commcare/utils/FileUtil.java [hunk 2]
```diff
@@ -21,6 +22,7 @@
 import android.widget.Toast;
 
 import org.commcare.CommCareApplication;
+import org.commcare.dalvik.BuildConfig;
```
Comment (shubham1g5) [inline app/src/org/commcare/utils/FileUtil.java:25]: can be removed now.

Comment (hashamyounis9) [reply app/src/org/commcare/utils/FileUtil.java]: done 9548872fc9eece60216d0948f0a24cc0ad56fb9a

**Review by shubham1g5 — APPROVED** [review]

**Comment by shubham1g5** [comment]
@hashamyounis9 we just merged an unrelated fix for the test failure that's causing the job to fail here, would you be able to rebase your branch from our `master` so that we can get the build passing here.

**Comment by hashamyounis9** [comment]
@shubham1g5 

rebase done with master in f1208218403fa6106fe9fd8e3309ab918aa180d2, please let me know if it resolves the issue.

**Comment by hashamyounis9** [comment]
@shubham1g5 

Thanks for your guidance and support along the way! ❤️
